File: 1094-car-pooling/1094-car-pooling.c


bool carPooling(int** trips, int tripsSize, int* tripsColSize, int capacity){
    int maxD=0;
    for(int i=0;i<tripsSize;i++) if(trips[i][2]>maxD) maxD=trips[i][2];
    
    int* pass=calloc(maxD+1,sizeof(int));
    for(int i=0;i<tripsSize;i++){
        pass[trips[i][1]]+=trips[i][0];
        pass[trips[i][2]]-=trips[i][0];
    }
    for(int p=0;p<maxD;p++){
        if(p-1>=0) pass[p]+=pass[p-1];
        if(pass[p]>capacity) return false;
    }
    return true;
        
}

File: 1094-car-pooling/1094-car-pooling.py
class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        n, MAX_DIST = len(trips), 1001
        passengers = [0]*MAX_DIST

        for pass_count, origin, dest in trips:
            passengers[origin] += pass_count
            passengers[dest] -= pass_count

        for i in range(1, MAX_DIST):
            passengers[i] += passengers[i-1]

        for pass_count in passengers:
            if pass_count > capacity:
                return False
        return True